src/common/k8s.py

```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Limited
+# See LICENSE file for licensing details.
+
+"""Common classes/functions for K8s implementations."""
+
+import logging
+from abc import ABC
+
+from ops import Container
+from ops.pebble import ExecError
+from typing_extensions import override
+
+from common.workload import AbstractWorkload
+
+logger = logging.getLogger(__name__)
+
+
+class K8sWorkload(AbstractWorkload, ABC):
+    """Class for providing implementation for IO operations on K8s."""
+
+    container: Container
+
+    def exists(self, path: str) -> bool:
+        """Check for file existence.
+
+        Args:
+            path: the full filepath to be checked for
+        """
+        return self.container.exists(path)
+
+    @override
+    def read(self, path: str) -> list[str]:
+        """Read from a file.
+
+        Args:
+            path: the full filepath to be read
+
+        Returns:
+            content of the file
+        """
+        if not self.container.exists(path):
```

Comment:
Uhm, I agree. An empty file is different from a not existing file. However, I did it in this way as 'm kinda extending the Workload class, and some implementation where already provided in Kafka and ZooKeeper (see [here](https://github.com/canonical/kafka-operator/blob/main/src/workload.py#L58)). 

Should we prescriptive to have charms with same behaviour for this? Maybe I would say so (given it would be good to have consistent `read` and `write` implementations), but in this case I would give +1 to @welpaolo 

@marcoppenheimer @juditnovak

Reply:
IMO whatever we choose to return is up to us.
Personally, I like empty lists as they're falsey, and avoids me having to `try: except` all the time, when really all I care about is if there is 'content' there or not, but I wouldn't want to impose that on other teams if they feel differently!

Reply:
My suggestion here would be to have a clean signature (that raises exceptions, given non-existence IS a different case of having a file empty) in the "general" class, that applies to all charms. If we don't want to be prescriptive, in the product specific implementation one can override this by wrapping the try except. 

Actually, I believe it is a good decorator use case to wrap try expect (with some default value to return when this happens), something like:

```
@safe(default=[])
def function_that_can_fail(....):
```

I believe I have implemented this previously in my life and I found it quite useful, for similar reason as Marc said. But I would not make it generic.